lib/galaxy/model/__init__.py

```diff
@@ -1142,6 +1172,75 @@ def command_version(self):
         for dataset_assoc in self.output_datasets:
             return dataset_assoc.dataset.tool_version
 
+    def update_output_states(self):
+        statements = ['''
+            UPDATE history_dataset_collection_association
+            SET update_time = :update_time
+            WHERE id in (
+                SELECT hdca.id
+                FROM history_dataset_collection_association hdca
+                INNER JOIN implicit_collection_jobs icj
+                    on icj.id = hdca.implicit_collection_jobs_id
+                INNER JOIN implicit_collection_jobs_job_association icjja
+                    on icj.id = icjja.implicit_collection_jobs_id
+                WHERE icjja.job_id = :job_id
+                UNION
+                SELECT hdca2.id
+                FROM history_dataset_collection_association hdca2
+                WHERE hdca2.job_id = :job_id
+            );
+        ''', '''
+            UPDATE dataset
+            SET
+                state = :state,
+                update_time = :update_time
+            WHERE id IN (
+                SELECT hda.dataset_id FROM history_dataset_association hda
+                INNER JOIN job_to_output_dataset jtod
+                ON jtod.dataset_id = hda.id AND jtod.job_id = :job_id
+            );
+        ''', '''
+            UPDATE dataset
+            SET
+                state = :state,
+                update_time = :update_time
+            WHERE id IN (
+                SELECT ldda.dataset_id FROM library_dataset_dataset_association ldda
+                INNER JOIN job_to_output_library_dataset jtold
+                ON jtold.ldda_id = ldda.id AND jtold.job_id = :job_id
+            );
+        ''', '''
+            UPDATE history_dataset_association
+            SET
```

Comment:
If performance is your primary concern you might consider updating from joins rather than with where-in subqueries. Traditionally that has been a more performant way to do this in many databases, though I do find it harder to understand visually, and I think the way you've written it is easier to understand for future coders.

If you choose to go that route, because of the variety of sql syntaxes for an update-from, this is a reasonably good place to leverage the orm's query builder funcs.

Reply:
`update from` is not supported in sqlite, and the code needs to support both postgres and sqlite. Checking the current db engine and executing the appropriate query is easy (here's one example https://github.com/galaxyproject/galaxy/pull/9515/files#diff-5da75d9743ecb1b3e59a6ba6527ccb16R63). My concern is that this module is already 6000+ lines long. It's bad enough it contains raw SQL.. Unless we *know* that using a join for postgres updates does indeed address a bottleneck, then sure, let's have 2 versions and execute them conditionally. Other than that, I'd keep it simple.